validator/sawtooth_validator/execution/scheduler_serial.py

```diff
@@ -154,30 +154,48 @@ def _get_batch_result(self, txn_id):
     def _dep_is_known(self, txn_id):
         return txn_id in self._txn_to_batch
 
-    def _dep_is_not_valid(self, txn_id):
-        dependency_result = self._get_batch_result(txn_id)
-        return not dependency_result.is_valid
+    def _in_invalid_batch(self, txn_id):
+        if self._txn_to_batch[txn_id] in self._batch_statuses:
+            dependency_result = self._get_batch_result(txn_id)
+            return not dependency_result.is_valid
+        return False
+
+    def _handle_fail_fast(self, txn):
+        self._set_batch_result(
+            txn.header_signature,
+            False,
+            None)
+
+        if txn.header_signature in self._last_in_batch:
+            self._previous_context_id = self._previous_valid_batch_c_id
 
     def next_transaction(self):
         with self._condition:
             if self._in_progress_transaction is not None:
                 return None
 
             txn = None
-            try:
-                while txn is None:
+            while txn is None:
+                try:
                     txn = self._txn_queue.get(block=False)
-                    # Handle this transaction being invalid based on a
-                    # dependency.
-                    if any(self._dep_is_known(d) and self._dep_is_not_valid(d)
-                            for d in self._get_dependencies(txn)):
-                        self._set_batch_result(
-                            txn.header_signature,
-                            False,
-                            None)
-                        txn = None
-            except queue.Empty:
-                return None
+                except queue.Empty:
+                    if self.complete(block=False):
+                        raise StopIteration()
+                    return None
+                # Handle this transaction being invalid based on a
+                # dependency.
+                if any(self._dep_is_known(d) and self._in_invalid_batch(d)
+                        for d in self._get_dependencies(txn)):
+                    self._set_batch_result(
+                        txn.header_signature,
+                        False,
+                        None)
+                    txn = None
+                    continue
```

Comment:
Will this continue with transactions that are still in the same batch?

Reply:
Yeah, the transactions are read in order from the queue, and if the batch has already been determined to be invalid, they are handled as invalid.